File: tsfile/src/main/java/org/apache/tsfile/file/metadata/TableSchema.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.tsfile.file.metadata;

import org.apache.tsfile.compatibility.DeserializeContext;
import org.apache.tsfile.utils.ReadWriteIOUtils;
import org.apache.tsfile.write.record.Tablet.ColumnType;
import org.apache.tsfile.write.schema.MeasurementSchema;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TableSchema {
  // the tableName is not serialized since the TableSchema is always stored in a Map, from whose
  // key the tableName can be known
  protected String tableName;
  protected List<MeasurementSchema> columnSchemas;
  protected List<ColumnType> columnTypes;
  protected boolean updatable = false;

  // columnName -> pos in columnSchemas;
  private Map<String, Integer> columnPosIndex;

  public TableSchema(String tableName) {
    this.tableName = tableName;
    this.columnSchemas = new ArrayList<>();
    this.columnTypes = new ArrayList<>();
    this.updatable = true;
  }

  public TableSchema(
      String tableName, List<MeasurementSchema> columnSchemas, List<ColumnType> columnTypes) {
    this.tableName = tableName;
    this.columnSchemas = columnSchemas;
    this.columnTypes = columnTypes;
  }

  public Map<String, Integer> getColumnPosIndex() {
    if (columnPosIndex == null) {
      columnPosIndex = new HashMap<>();
    }
    return columnPosIndex;
  }

  public int findColumnIndex(String columnName) {
    return getColumnPosIndex()
        .computeIfAbsent(
            columnName,
            colName -> {
              for (int i = 0; i < columnSchemas.size(); i++) {
                if (columnSchemas.get(i).getMeasurementId().equals(columnName)) {
                  return i;
                }
              }
              return -1;
            });
  }

  public MeasurementSchema findColumnSchema(String columnName) {
    final int columnIndex = findColumnIndex(columnName);
    return columnIndex >= 0 ? columnSchemas.get(columnIndex) : null;
  }

  public void update(ChunkGroupMetadata chunkGroupMetadata) {
    if (!updatable) {
      return;
    }

    for (ChunkMetadata chunkMetadata : chunkGroupMetadata.getChunkMetadataList()) {
      int columnIndex = findColumnIndex(chunkMetadata.getMeasurementUid());
      // if the measurement is not found in the column list, add it
      if (columnIndex == -1) {
        columnSchemas.add(chunkMetadata.toMeasurementSchema());
        columnTypes.add(ColumnType.MEASUREMENT);
        getColumnPosIndex().put(chunkMetadata.getMeasurementUid(), columnSchemas.size() - 1);
      }
    }
  }

  public List<MeasurementSchema> getColumnSchemas() {
    return columnSchemas;
  }

  public List<ColumnType> getColumnTypes() {
    return columnTypes;
  }

  public int serialize(OutputStream out) throws IOException {
    int cnt = 0;
    if (columnSchemas != null) {
      cnt += ReadWriteIOUtils.write(columnSchemas.size(), out);
      for (int i = 0; i < columnSchemas.size(); i++) {
        MeasurementSchema columnSchema = columnSchemas.get(i);
        ColumnType columnType = columnTypes.get(i);
        cnt += columnSchema.serializeTo(out);
        cnt += ReadWriteIOUtils.write(columnType.ordinal(), out);
      }
    } else {
      cnt += ReadWriteIOUtils.write(0, out);
    }

    return cnt;
  }

  public static TableSchema deserialize(ByteBuffer buffer, DeserializeContext context) {
    final int tableNum = buffer.getInt();
    List<MeasurementSchema> measurementSchemas = new ArrayList<>(tableNum);
    List<ColumnType> columnTypes = new ArrayList<>();
    for (int i = 0; i < tableNum; i++) {
      MeasurementSchema measurementSchema =
          context.measurementSchemaDeserializer.deserialize(buffer, context);
      measurementSchemas.add(measurementSchema);
      columnTypes.add(ColumnType.values()[buffer.getInt()]);
    }
    return new TableSchema(null, measurementSchemas, columnTypes);
  }

  public String getTableName() {
    return tableName;
  }

  public void setTableName(String tableName) {
    this.tableName = tableName;
  }
}

File: csharp/src/Apache.TsFile/Schema/TableSchema.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using Apache.TsFile.Enums;

namespace Apache.TsFile.Schema;

/// <summary>
/// Defines the schema for a table containing multiple measurements.
/// </summary>
public class TableSchema
{
    /// <summary>
    /// Gets the list of measurement schemas in this table.
    /// </summary>
    public List<MeasurementSchema> Measurements { get; }
    
    /// <summary>
    /// Gets the list of column schemas (v4 format) in this table.
    /// </summary>
    public List<ColumnSchema>? ColumnSchemas { get; set; }
    
    /// <summary>
    /// Gets or sets the table name.
    /// </summary>
    public string TableName { get; set; }

    /// <summary>
    /// Gets or sets whether this is a logical table schema (auto-generated from tree model).
    /// </summary>
    public bool IsLogicalTable { get; set; }

    /// <summary>
    /// Gets or sets the maximum level for tree model ID columns.
    /// </summary>
    public int MaxIdLevel { get; set; }
    
    /// <summary>
    /// Initializes a new instance of the TableSchema class.
    /// </summary>
    public TableSchema(string tableName)
    {
        if (string.IsNullOrWhiteSpace(tableName))
            throw new ArgumentException("Table name cannot be null or empty", nameof(tableName));
        
        TableName = tableName;
        Measurements = new List<MeasurementSchema>();
    }
    
    /// <summary>
    /// Adds a measurement schema to this table.
    /// </summary>
    public void AddMeasurement(MeasurementSchema measurement)
    {
        if (measurement == null)
            throw new ArgumentNullException(nameof(measurement));
        
        if (Measurements.Any(m => m.MeasurementName == measurement.MeasurementName))
            throw new ArgumentException($"Measurement {measurement.MeasurementName} already exists in table");
        
        Measurements.Add(measurement);
    }
    
    /// <summary>
    /// Gets a measurement schema by name.
    /// </summary>
    public MeasurementSchema? GetMeasurement(string measurementName)
    {
        return Measurements.FirstOrDefault(m => m.MeasurementName == measurementName);
    }
    
    /// <summary>
    /// Gets the number of measurements in this table.
    /// </summary>
    public int MeasurementCount => Measurements.Count;
    
    /// <summary>
    /// Serializes the table schema to a binary stream.
    /// </summary>
    public void Serialize(BinaryWriter writer)
    {
        writer.Write(TableName);
        writer.Write(Measurements.Count);
        
        foreach (var measurement in Measurements)
        {
            measurement.Serialize(writer);
        }
    }
    
    /// <summary>
    /// Deserializes the table schema from a binary stream.
    /// </summary>
    public static TableSchema Deserialize(BinaryReader reader)
    {
        var tableName = reader.ReadString();
        var count = reader.ReadInt32();
        
        var tableSchema = new TableSchema(tableName);
        
        for (int i = 0; i < count; i++)
        {
            var measurement = MeasurementSchema.Deserialize(reader);
            tableSchema.AddMeasurement(measurement);
        }
        
        return tableSchema;
    }
    
    /// <summary>
    /// Deserializes the table schema from a binary stream (v4 format).
    /// </summary>
    public static TableSchema DeserializeV4(string tableName, BinaryReader reader, 
        Func<int> readVarInt, Func<string> readVarIntString)
    {
        var tableSchema = new TableSchema(tableName);
        
        // Read column schemas
        var columnCount = readVarInt();
        tableSchema.ColumnSchemas = new List<ColumnSchema>();
        
        for (int i = 0; i < columnCount; i++)
        {
            var columnSchema = ColumnSchema.Deserialize(reader, readVarInt, readVarIntString);
            tableSchema.ColumnSchemas.Add(columnSchema);
            
            // Convert FIELD columns to MeasurementSchema for backward compatibility
            if (columnSchema.Category == ColumnCategory.Field)
            {
                var measurement = new MeasurementSchema(
                    columnSchema.Name,
                    columnSchema.DataType,
                    columnSchema.Encoding,
                    columnSchema.Compression
                );
                tableSchema.Measurements.Add(measurement);
            }
        }
        
        return tableSchema;
    }
    
    public override string ToString()
    {
        return $"TableSchema{{Name={TableName}, Measurements={Measurements.Count}}}";
    }

    /// <summary>
    /// Creates a logical table schema from tree model device path.
    /// </summary>
    /// <param name="devicePath">Device path like "root.db1.d1"</param>
    /// <param name="measurements">Measurement schemas</param>
    /// <param name="segmentNumForTableName">Number of segments for table name (default 3)</param>
    public static TableSchema CreateFromTreeModel(string devicePath, List<MeasurementSchema> measurements,
        int segmentNumForTableName = 3)
    {
        var segments = devicePath.Split('.');

        // For tree model, use the full device path as table name
        // This ensures each device has its own logical table
        var schema = new TableSchema(devicePath)
        {
            IsLogicalTable = true,
            MaxIdLevel = Math.Max(1, segments.Length - segmentNumForTableName + 1)
        };

        foreach (var m in measurements)
        {
            schema.AddMeasurement(m);
        }

        return schema;
    }
}
